### PR TITLE
image,c/snap: implement prepare-image --customize

### PR DESCRIPTION
**Pull request by pedronis**
this moves special customization responsabilities from ubuntu-image to prepare-image

this would allow to support them on UC20 with grade dangerous if requested and reduces
the u-i work to support more of these if needed

this matches https://github.com/CanonicalLtd/ubuntu-image/pull/199/ (I tested the two combined to some extent)

### CODE REVIEW

File: cmd/snap/cmd_prepare_image.go [hunk 1]
```diff
@@ -92,6 +99,19 @@ func (x *cmdPrepareImage) Execute(args []string) error {
 		Architecture: x.Architecture,
 	}
 
+	if x.Customize != "" {
+		f, err := os.Open(x.Customize)
+		if err != nil {
+			return fmt.Errorf("cannot read image customizations: %v", err)
+		}
```
Comment (mvo5) [inline cmd/snap/cmd_prepare_image.go:106]: Do we need a `defer f.Close()` here or is something else closing this file?

File: cmd/snap/cmd_run.go [hunk 2]
```diff
@@ -111,7 +111,7 @@ and environment.
 			// TRANSLATORS: This should not start with a lowercase letter.
 			"gdb": i18n.G("Run the command with gdb (deprecated, use --gdbserver instead)"),
 			// TRANSLATORS: This should not start with a lowercase letter.
-			"gdbserver": i18n.G("Run the command with gdbserver"),
+			"gdbserver":              i18n.G("Run the command with gdbserver"),
```
Comment (mvo5) [inline cmd/snap/cmd_run.go:114]: Did this land on purpose?

**Review by mvo5 — APPROVED** [review]
Very nice! Two questions inline but once resolved this is good to go I think.

File: cmd/snap/cmd_prepare_image.go (hunk 1 above)
Comment (pedronis) [reply cmd/snap/cmd_prepare_image.go:106]: no, thanks for spotting, we need it, but probably the code needs a reorg, is a bit weird to keep that file open until we have created the image

File: cmd/snap/cmd_run.go (hunk 2 above)
Comment (pedronis) [reply cmd/snap/cmd_run.go:114]: not sure, I missed I was checking this in

File: cmd/snap/cmd_prepare_image.go (hunk 1 above)
Comment (pedronis) [reply cmd/snap/cmd_prepare_image.go:106]: done

**Review by sil2100 — APPROVED** [review]
I can only do an overall code/concept review here, but it all seems to look good! Thanks! I see that for now this would be still disabled for UC20 - which is okay, since that was the case in ubuntu-image right now. But I think it would be nice if we could get some customization enabled for UC20 cases in the near future as well.

Anyway, this looks like what we want.

File: image/image_linux.go [hunk 3]
```diff
@@ -144,6 +171,36 @@ func installCloudConfig(rootDir, gadgetDir string) error {
 	return osutil.CopyFile(cloudConfig, dst, osutil.CopyFlagOverwrite)
 }
 
+func customizeImage(rootDir, defaultsDir string, custo *Customizations) error {
+	// customize with cloud-init user-data
+	if custo.CloudInitUserData != "" {
+		varCloudDir := filepath.Join(rootDir, "/var/lib/cloud/seed/nocloud-net")
```
Comment (bboozzoo) [inline image/image_linux.go:177]: It'd be great to include a reference to those doc pages in the comment:
https://cloudinit.readthedocs.io/en/latest/topics/dir_layout.html
https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html

File: image/image_linux.go [hunk 4]
```diff
@@ -144,6 +171,36 @@ func installCloudConfig(rootDir, gadgetDir string) error {
...
+		if err := os.MkdirAll(varCloudDir, 0755); err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(filepath.Join(varCloudDir, "meta-data"), []byte("instance-id: nocloud-static\n"), 0644); err != nil {
+			return err
+		}
+		dst := filepath.Join(varCloudDir, "user-data")
+		if err := osutil.CopyFile(custo.CloudInitUserData, dst, osutil.CopyFlagOverwrite); err != nil {
```
Comment (bboozzoo) [inline image/image_linux.go:185]: Should we verify that user-data is at least valid YAML? If not now then maybe a TODO would be applicable.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: image/image_linux.go (hunk 4 above)
Comment (pedronis) [reply image/image_linux.go:185]: u-i image itself wasn't doing that check, this is kind of a legacy feature at this point

File: image/image_linux.go (hunk 3 above)
Comment (pedronis) [reply image/image_linux.go:177]: I will add this in a follow up

**Pull request merged**